components/pkg-export-tar/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::fs::File;
+pub use cli::{Cli, PkgIdentArgOptions};
+pub use error::{Error, Result};
+use common::ui::UI;
+use hcore::channel;
+use hcore::url as hurl;
+use hcore::package::{PackageIdent, PackageArchive};
+use mktemp::Temp;
+use regex::Regex;
+use tar::Builder;
+
+pub use build::BuildSpec;
+
+/// The version of this library and program when built.
+pub const VERSION: &'static str = include_str!(concat!(env!("OUT_DIR"), "/VERSION"));
+
+pub fn export_for_cli_matches(ui: &mut UI, matches: &clap::ArgMatches) -> Result<()> {
+    let default_channel = channel::default();
+    let default_url = hurl::default_bldr_url();
+    let spec = BuildSpec::new_from_cli_matches(&matches, &default_channel, &default_url);
+    export(ui, spec)?;
+
+    Ok(())
+}
+
+pub fn export(ui: &mut UI, build_spec: BuildSpec) -> Result<()> {
+
+    let hart_to_package = build_spec.idents_or_archives.join(", ");
+    let builder_url = build_spec.url;
+
+    ui.begin(
+        format!("Building a tarball with: {}", hart_to_package),
+    )?;
+
+    let temp_dir_path = Temp::new_dir().unwrap().to_path_buf();
+
+    initiate_tar_command(&temp_dir_path, &hart_to_package, &builder_url);
+
+    Ok(())
+}
+
+fn initiate_tar_command(temp_dir_path: &PathBuf, hart_to_package: &str, builder_url: &str) {
+    let status = Command::new("hab")
+        .arg("studio")
+        .arg("-r")
+        .arg(&temp_dir_path)
+        .arg("new")
+        .status()
+        .expect("failed to create studio");
+
+    if status.success() {
+        println!("Able to create studio to export package as tarball, proceeding...");
+        install_command(&temp_dir_path, &hart_to_package, &builder_url);
+    } else {
+        println!("Unable to create a studio to export the package as a tarball.")
+    }
+}
+
+fn install_command(temp_dir_path: &PathBuf, hart_to_package: &str, builder_url: &str) {
+    let status = Command::new("hab")
+        .arg("studio")
+        .arg("-q")
+        .arg("-r")
+        .arg(&temp_dir_path)
+        .arg("run")
+        .arg("hab")
+        .arg("install")
+        .arg("-u")
+        .arg(builder_url)
+        .arg(&hart_to_package)
+        .status()
+        .expect("failed to install package in studio");
+
+    if status.success() {
+        println!("Hart package is installable in a studio, exporting tarball...");
+        tar_command(&temp_dir_path, &hart_to_package);
+    } else {
+        println!("Hart package is NOT installable in a studio and could not be exported.");
+        println!("Please see the above error for details.");
+    }
+}
+
+fn tar_command(temp_dir_path: &PathBuf, hart_to_package: &str) {
+    let tarball = File::create(tar_name(&hart_to_package)).unwrap();
+    let mut tar_builder = Builder::new(tarball);
+
+    let mut hab_pkgs_buf = temp_dir_path.clone();
+    hab_pkgs_buf.push("hab/pkgs");
+
+    tar_builder.append_dir_all("hab/pkgs", hab_pkgs_buf);
```

Comment:
How does this tar crate deal with Unix vs. Windows paths? Does it do any internal translation of path separators?

Reply:
Windows can handle `/` in paths. So this should work fine. The times that causes problems are when application logic does not expect `/` which is rare and likely not the case here. That said, we should for sure give this a test run on a windows package on windows for sanity.

Reply:
@mwrock Good to know... sounds like I need to update my Windows knowledge from the Windows98 era 😄